Request-level circuit breaker support on coordinating nodes

jimczi opened the pull request:
This commit allows coordinating node to account the memory used to perform partial and final reduce of
aggregations in the request circuit breaker. The search coordinator adds the memory that it used to save
and reduce the results of shard aggregations in the request circuit breaker. Before any partial or final
reduce, the memory needed to reduce the aggregations is estimated and a CircuitBreakingException is thrown
if exceeds the maximum memory allowed in this breaker.
This size is estimated as roughly 1.5 times the size of the serialized aggregations that need to be reduced.
This estimation can be completely off for some aggregations but it is corrected with the real size after
the reduce completes.
If the reduce is successful, we update the circuit breaker to remove the size of the source aggregations
and replace the estimation with the serialized size of the newly reduced result.

As a follow up we could trigger partial reduces based on the memory accounted in the circuit breaker instead
of relying on a static number of shard responses. A simpler follow up that could be done in the mean time is
to [reduce the default batch reduce size](https://github.com/elastic/elasticsearch/issues/51857) of blocking
search request to a more sane number.

Closes #37182

jimczi committed:
Request-level circuit breaker support on coordinating nodes

This commit allows coordinating node to account the memory used to perform partial and final reduce of
aggregations in the request circuit breaker. The search coordinator adds the memory that it used to save
and reduce the results of shard aggregations in the request circuit breaker. Before any partial or final
reduce, the memory needed to reduce the aggregations is estimated and a CircuitBreakingException} is thrown
if exceeds the maximum memory allowed in this breaker.
This size is estimated as roughly 1.5 times the size of the serialized aggregations that need to be reduced.
This estimation can be completely off for some aggregations but it is corrected with the real size after
the reduce completes.
If the reduce is successful, we update the circuit breaker to remove the size of the source aggregations
and replace the estimation with the serialized size of the newly reduced result.

As a follow up we could trigger partial reduces based on the memory accounted in the circuit breaker instead
of relying on a static number of shard responses. A simpler follow up that could be done in the mean time is
to [reduce the default batch reduce size](https://github.com/elastic/elasticsearch/issues/51857) of blocking
search request to a more sane number.

Closes #37182

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Circuit Breakers)

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Geo)

jimczi commented on server/src/test/java/org/elasticsearch/action/search/TransportSearchActionSingleNodeTests.java:
```diff
@@ -1,177 +0,0 @@
-/*
- * Licensed to Elasticsearch under one or more contributor
```
Note for reviewer: I moved these tests in https://github.com/elastic/elasticsearch/pull/62223/files#diff-0721d29fdc234c0f88a9019057ea55bd

javanna replied:
it is a bit of a shame that these go from single node tests to full blown IT tests, what is the reasoning behind this choice?

jimczi replied:
I regrouped the search action tests in a single IT class. I agree that these tests may not require the full IT but they are grouped with other tests that require it so I thought that it makes sense to move them here.

jimczi committed:
style

jimczi committed:
Merge branch 'master' into enhancements/reduce_aggs_circuit_breaker

jimczi committed:
style

jimczi committed:
fix unit test

nik9000 reviewed (commented):
This looks right to me. I'll have to go over it more closely before  :+1: it and I think I'd like to wait a day just to have fresh eyes on it.

jimczi committed:
Merge branch 'master' into enhancements/reduce_aggs_circuit_breaker

jimczi committed:
Don't serialize partial reduce eagerly

This commit removes the serialization of partial reduce in order to speed up the merges when
the batch reduce size is smaller than the number of shards in the request.
The estimation of the size of partial reduce is still based on the binary size (serialized form) but we
keep the full java object and estimate the size with a counting stream output.
Finally this change adds a benchmark for the reduce of nested terms aggs. This benchmark was used to
optimize the code in this PR.

jimczi committed:
add a seed for the aggs benchmark

jimczi committed:
fix ut

jimczi commented:
We discussed offline with @nik9000 and I pushed some changes to speed up the partial merge. First of all I removed the serialization of the partial reduce and replace it with an estimation of the size based on a noop-serialization (just counting the bytes). That resulted in much better performance for large cardinality aggregations and allows to estimate more precisely the memory used by the final reduce. I added the benchmark that I used in the PR to be able to replay the numbers.
For instance a terms/terms agg with a topN size of 100, 512 shards and an overall cardinality of 10,000 gives the following numbers on my machine:
````
Benchmark                        (bufferSize)  (cardinalityFactor)  (numShards)  (topNSize)  Mode  Cnt      Score      Error  Units
TermsReduceBenchmark.reduceAggs           512                  100         512         100  avgt    7   1756,207 ±   15,345  ms/op
TermsReduceBenchmark.reduceAggs            32                  100          512         100  avgt    7  27112,587 ± 13811,555  ms/op
TermsReduceBenchmark.reduceAggs            32                  100          512         100  avgt    7  10336,829 ± 2762,412  ms/op
````
The first result is the time it takes to reduce with a batch reduce size of 512, the second result is when we serialize the results of partial aggs with a batch reduce size of 32 and the last one is when we don't serialize partial results. As you can see the speedups are significant.

jimczi committed:
Merge branch 'master' into enhancements/reduce_aggs_circuit_breaker

jimczi committed:
adapt new code after master merge

nik9000 commented on server/src/internalClusterTest/java/org/elasticsearch/action/search/TransportSearchIT.java:
```diff
@@ -19,22 +19,242 @@
 
 public class TransportSearchIT extends ESIntegTestCase {
+    public static class TestPlugin extends Plugin implements SearchPlugin {
+        public TestPlugin(Settings settings) {}
```
I believe plugins don't *need* to take a `Settings` as an argument.

jimczi replied:
++, thanks

nik9000 commented on server/src/internalClusterTest/java/org/elasticsearch/action/search/TransportSearchIT.java:
```diff
@@ -104,4 +324,274 @@ public void testSearchIdle() throws Exception {
+
+    public void testCircuitBreakerReduceFail() throws Exception {
+        int numShards = randomIntBetween(1, 10);
+        indexSomeDocs("test", numShards, numShards*3);
```
I love the name of this method. It makes me think "I absolutely don't care about the contents of these docs." They are just "some docs".

nik9000 commented on server/src/internalClusterTest/java/org/elasticsearch/action/search/TransportSearchIT.java:
```diff
@@ -104,4 +324,274 @@ public void testSearchIdle() throws Exception {
+
+        @Override
+        public void postCollection() throws IOException {}
+    }
```
All the new tests make me happy!

nik9000 commented on server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java:
```diff
@@ -193,49 +206,106 @@ private MergeResult partialReduce(MergeTask task,
+                return 0;
+            }
+            return result.aggregations()
+                .asSerialized(InternalAggregations::readFrom, namedWriteableRegistry)
```
The `aggregations` should always be serialized here, right? I wonder if it is worth asserting that or something.

nik9000 commented on server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java:
```diff
@@ -164,27 +182,22 @@ private MergeResult partialReduce(MergeTask task,
-            aggsMaxBufferSize = Math.max(aggsCurrentBufferSize, aggsMaxBufferSize);
-            logger.trace("aggs partial reduction [{}->{}] max [{}]",
-                previousBufferSize, aggsCurrentBufferSize, aggsMaxBufferSize);
+            newAggs = result;
```
I think you can skip the `result` variable now.

jimczi replied:
++, https://github.com/elastic/elasticsearch/pull/62223/commits/7178dab4566971b0b0daee19f2b22a8f597f61b1

nik9000 commented on server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java:
```diff
@@ -193,49 +206,106 @@ private MergeResult partialReduce(MergeTask task,
         }
         progressListener.notifyPartialReduce(processedShards, topDocsStats.getTotalHits(), newAggs, numReducePhases);
-        return new MergeResult(processedShards, newTopDocs, newAggs);
+        return new MergeResult(processedShards, newTopDocs, newAggs, hasAggs ? newAggs.getBinarySize() : 0);
```
Its probably worth commenting about how we leave the results un-serialized but we send their serialized size because serializing is slow.

jimczi replied:
https://github.com/elastic/elasticsearch/pull/62223/commits/7178dab4566971b0b0daee19f2b22a8f597f61b1

nik9000 commented on server/src/main/java/org/elasticsearch/search/aggregations/InternalAggregations.java:
```diff
@@ -166,4 +167,47 @@ public static InternalAggregations reduce(List<InternalAggregations> aggregation
+    /**
+     * Returns the number of bytes required to serialize these aggregations in binary form.
+     */
+    public long getBinarySize() {
```
`getSerializedSize`?

jimczi replied:
https://github.com/elastic/elasticsearch/pull/62223/commits/7178dab4566971b0b0daee19f2b22a8f597f61b1

nik9000 commented on server/src/test/java/org/elasticsearch/action/search/SearchPhaseControllerTests.java:
```diff
@@ -931,19 +935,36 @@ public void onFinalReduce(List<SearchShard> shards, TotalHits totalHits, Interna
 
-    public void testPartialMergeFailure() throws InterruptedException {
+    public void testPartialReduce() throws Exception {
+        for (int i = 0; i < 10; i++) {
```
Are these loops just to make sure we get a whole bunch of random cases every run?

jimczi replied:
yes,  mainly to exercice the batch reduced size

nik9000 reviewed (approved):
I left some small things but LGTM.

It's a shame about reserializing being slow.

jimczi committed:
Merge branch 'master' into enhancements/reduce_aggs_circuit_breaker

jimczi committed:
address review feedback

jimczi committed:
asserting doesn't work with tests, revert. Fix tests after merge with master.

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java:
```diff
@@ -401,16 +400,15 @@ public void onPartialReduce(List<SearchShard> shards, TotalHits totalHits,
                  * *already* has that reference so we're not creating more garbage.
                  */
                 reducedAggs = () ->
-                    InternalAggregations.topLevelReduce(singletonList(aggregations.expand()), aggReduceContextSupplier.get());
```
what is the rationale behind the choice of keeping the de-serialized partial aggs around rather then serializing them and expanding them later when needed?

jimczi replied:
Because serializing and de-serializing is slow and makes the memory consumption worst when expanding. We can maybe revise this later but I feel like the real difference would be to serialize on a temporary file or in an index in case of async_search.

javanna replied:
I see, so this particular change does not really have to do with circuit breaking, but more around the feeling that keeping aggs around as they are is better overall? Was this measured out of curiosity? I think I am asking because I wonder why we were serializing/deserializing before

jimczi replied:
> Was this measured out of curiosity? 

Yes, see https://github.com/elastic/elasticsearch/pull/62223#issuecomment-692928324

javanna replied:
++ great thanks

javanna replied:
one thing that could have made things clearer would be to have two separate PRs for the two changes, but it's fine as-is at this point, and I am maybe missing some reason why they were made as part of a single PR.

javanna commented on x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchTaskTests.java:
```diff
@@ -155,56 +152,14 @@ public void onFailure(Exception e) {
-        assertThat(response.get().getFailure().getCause(), instanceOf(IllegalArgumentException.class));
-        assertEquals("Unknown NamedWriteable category [" + InternalAggregation.class.getName() + "]",
-            response.get().getFailure().getCause().getMessage());
-    }
```
is this test no longer relevant?

jimczi replied:
it cannot work anymore since we don't need to serialize the aggs. I think it's ok since we have other tests that check that exception thrown during a partial/final reduce are handled correctly.

javanna replied:
I see, you mean the condition that the test had to trigger the failure, which was around serialization?

jimczi replied:
yep

javanna reviewed (commented):
left a couple of small questions

nik9000 commented:
Sorry not to comment publicly about the serialization change! It makes me sad not to serialize but I see the reasoning.

jimczi committed:
Merge branch 'master' into enhancements/reduce_aggs_circuit_breaker

jimczi committed:
fix compilation after merge with  master

jimczi commented:
@elasticmachine run elasticsearch-ci/2
